### DISCUSSION TITLE
Use Gradle build tool instead of SBT

### DISCUSSION THREAD

**Discussion by cgpoh** [unitycatalog/unitycatalog#218] [Ideas]
Currently, the only way to build the project is by using SBT. Consider adding support to use Gradle for building the project.

**Comment by havryliuk** [comment]
Why do we need to add gradle instead of sbt?
We would have to support both tools then.

**Reply by creechy** [reply]
The OP says "add" but that wouldn't be smart to have multiple ways to build. If anything it would be replace sbt with gradle.

**Reply by MrPowers** [reply]
@creechy - good call out.  I updated the Discussion title to reflect that this discussion is about replacing SBT with Gradle.

**Comment by cgpoh** [comment]
Move the build tools to Gradle? My colleague having problem with sbt

**Comment by havryliuk** [comment]
I don't know, I am not an owner of the project, I guess they selected it for some reason. Don't see why it should be changed.

**Comment by MrPowers** [comment]
Gradle is a build tool that's popular in the Java ecosystem.  Some data projects like [Apache Iceberg](https://github.com/apache/iceberg) use Gradle.

SBT is popular in the Scala ecosystem but also works for Java projects.

Most users just fetch the code artifacts from Maven, so the build tool should be whatever works best for the devs.

I was chatting with @filipovskid, and he mentioned that SBT was difficult for him to use on Windows, but it worked well on a Mac.  I don't know if this is a general issue or a machine-specific issue.  This may be a factor we would like to consider if SBT is challenging for Windows devs.

**Reply by jaceklaskowski** [reply]
I'm a big fan of sbt and barely use gradle so I'm certainly biased towards more sbt whenever possible 😉 

But this project is mostly Java (with just a tiny bit of Scala ~3% and Python ~1%, cf. ) so it makes sense to use gradle more than sbt. If there's someone to take care of it, I'd support it.

p.s. On Windows even the simplest things are fairly difficult. Windows has never been very developer-friendly IMHO 🤣

**Reply by SemyonSinchenko** [reply]
> Gradle is a build tool that's popular in the Java ecosystem. Some data projects like [Apache Iceberg](https://github.com/apache/iceberg) use Gradle.
> 
> SBT is popular in the Scala ecosystem but also works for Java projects.
> 
> Most users just fetch the code artifacts from Maven, so the build tool should be whatever works best for the devs.
> 
> I was chatting with @filipovskid, and he mentioned that SBT was difficult for him to use on Windows, but it worked well on a Mac. I don't know if this is a general issue or a machine-specific issue. This may be a factor we would like to consider if SBT is challenging for Windows devs.

Another potential issue with sbt: I briefly checked and it looks like one of the most popular Java Language Servers (JDTLS) does not support sbt-builds out-of-the-box. Only [Maven and Gradle](https://github.com/eclipse-jdtls/eclipse.jdt.ls?tab=readme-ov-file#features) are mentioned in features. So, it maybe challenged to import the project in some text-editors that relies on JDTLS.

**Reply by filipovskid** [reply]
I am quite biased, but there is an undeniable point vouching for a transition: 

* Ease of use and contributor accessibility: Maven and Gradle are more widely adopted in the Java ecosystem, making them familiar to a larger pool of developers. This can lower the barrier to entry for new contributors, enhancing community involvement and speeding up the onboarding process. Besides this, these systems have extensive documentation, a rich repository of plugins for various tasks, and strong community support.

If the target contributors are developers with a Scala background, then it is reasonable to keep SBT, otherwise I would move away from this painful experience.

**Reply by SemyonSinchenko** [reply]
One of the solutions to marry sbt and Eclipse Java LSP (JDTLS) is to use [this sbt plugin](https://github.com/sbt/sbt-eclipse). It should simplify import to any IDEs that use JDTLS (NeoVim, Emacs, VSCode, etc.). LICENSE of the plugin fits with the project.

**Reply by ni-mi** [reply]
While Gradle is widely used - Maven is hands down the most popular build tool for Java users - and of course much more than SBT which is not used in Java projects.

[Here you can see in Google trends how much Maven is more popular ](https://trends.google.com/trends/explore?date=today%205-y&q=maven,gradle&hl=en)

SBT is more problematic to look for in Google trends since SBT has more meanings than just the Simple Build Tool - but there are many sites the say the same...

I believe that switching to Maven should be easy enough.

**Reply by jaceklaskowski** [reply]
Oh no, please no Maven. It's such an old build tool. I'd rather see gradle (that could be one of the many reasons to attract more Java devs to hop onto the project's bandwagon, just for the sake of learning this still new shiny build tool).

**Reply by creechy** [reply]
Agree. I resisted Maven for a long time, but I eventually did like the convention over configuration mentality. Even still Gradle seems more recently popular than Maven.

**Reply by cgpoh** [reply]
Yep, no maven please. We migrated our build tool from maven to gradle last year and with unity catalog using gradle, it will be easier for us to integrate into our build tool.

@MrPowers , thanks for updating the discussion title!

**Comment by ognis1205** [comment]
I think both sbt and gradle are good build tools, but in Java projects, using sbt instead of gradle is a bit offbeat from what I think were used to. Unless there is a specific reason, I think it is important to follow conventional development practices to attract developers. People are more likely to participate in projects that follow familiar conventions. This is more a matter of perception than practical utility.
